rolldown/rolldown
# feat(plugin/deno-loader): add draft

### Description
This PR adds a`DenoLoaderPlugin` to improve support for Deno-specific module resolution and loading. The main additions include:

1. Added support for resolving and loading modules from JSR (JavaScript Registry) using the `jsr:` protocol.
2. Added support of `npm:` specifiers using inherited resolution.
3. Added support for HTTP and HTTPS module loading.
4. Implemented a caching mechanism for Deno module information to improve performance.
5. Integrated import map parsing and resolution for better module specifier handling.

These changes allow the plugin to work with Deno's module system, including third-party modules from JSR and npm, as well as remote modules loaded via HTTP(S).

### Implementation Details

- Added a `DenoResolveResult` struct to cache and manage resolved module information.
- Implemented `get_deno_info` and `follow_redirects` functions to interact with Deno's module resolution system.
- Implemented the `resolve_id` method to handle various module specifier types (JSR, npm, HTTP(S)).
- Implemented the `load` method to support loading content for JSR and remote HTTP(S) modules using Deno Cache.
- Integrated import map parsing to resolve module specifiers according to Deno's import map specifications.

### Testing

To test this PR:

Check `rolldown/crates/rolldown_plugin_deno_loader/js_tests`

But for overview here the usage:
```ts
import {
  defineConfig,
  rolldown,
} from "../../../packages/rolldown/dist/esm/index.mjs";
import { denoLoaderPlugin } from "../../../../rolldown/packages/rolldown/dist/esm/experimental-index.mjs";

const config = defineConfig({
  input: { jsr: "./mod.ts" },
  resolve: { conditionNames: ["import"] },
  plugins: [
    denoLoaderPlugin({
      entryPoints: ["./mod.ts"],
      importMapBaseUrl: import.meta.resolve("./"),
      importMap: await fetch(import.meta.resolve("./deno.json"))
        .then((r) => r.text()),
    }),
  ],
});

await rolldown(config).then((build) => build.write(config.output));
```


### Additional Notes

This PR aims to add a native `DenoLoaderPlugin` to offer compatibility with Deno's module resolution system, allowing for a smoother usage of Deno with Rolldown projects.

Latest published package:

https://www.npmjs.com/package/@rolldown-deno-fork/rolldown/v/0.15.0-fork.1-snapshot-017e412-20241211232210
